### PR TITLE
Fixing all Examples

### PR DESCRIPTION
**Pull request by eduardolneto**
This is a pull request to clean examples and make them work with PR #36 to PR #40. Due to the new changed requesting during pull request some of them stopped working. So I had to add some changes to coils and surface classes to make them work again. There is however, still a problem with examples using stochastic optimization and the classifier object.

### CODE REVIEW

File: tests/test_objective_functions.py [hunk 1]
```diff
@@ -92,8 +136,9 @@ def setUp(self):
     @patch('essos.objective_functions.Curves', return_value=DummyCurves())
     @patch('essos.objective_functions.Coils', return_value=DummyCoils())
     @patch('essos.objective_functions.BiotSavart', return_value=DummyField())
-    @patch('essos.objective_functions.perturb_curves', side_effect=lambda curves, sampler, key=None, perturbation_type=None: curves)
-    def test_perturbed_field_and_coils_from_dofs(self, pc, bs, coils, curves):
+    @patch('essos.objective_functions.perturb_curves_systematic')
```
Comment (rogeriojorge) [inline tests/test_objective_functions.py]: Does this perturb_curves_systematic exist? I'm also checking in the objective_functions.py and it appears in line 206, but it is never imported. Doesn't this error out for you?

The same for perturb_curves_statistic, pertubred_field_from_dofs, field_from_dofs, and coils_from_dofs, as the current objective_functions.py only imports perturb_curves and has moved toward object-based losses

File: tests/test_objective_functions.py [hunk 2]
```diff
@@ -55,6 +56,49 @@ class DummySurface:
...
+        return self.gamma.shape[0]
+
+    def tree_flatten(self):
+        children = (self.gamma, self.gamma_dash, self.gamma_dashdash, self.currents, self.quadpoints)
+        return children, None
+
+    @classmethod
+    def tree_unflatten(cls, aux_data, children):
+        return cls(*children)
+
```
Comment (rogeriojorge) [inline tests/test_objective_functions.py:124]: Doesn't this error out for you? Seems like PytreeCoils is missing attributes used by the tested losses. loss_coil_surface_distance() needs coils.stellsym and coils.nfp; loss_linkingnumber() and loss_lorentz_force_coils() need coils.curves.quadpoints. What about using a real Coils object in these tests or add a small .curves.quadpoints, .stellsym, and .nfp to the dummy object?

File: examples/particle_tracing/trace_particles_coils_guidingcenter.py [hunk 3]
```diff
@@ -21,7 +21,7 @@
 energy=4000*ONE_EV
 
 # Load coils and field
-json_file = os.path.join(os.path.dirname(__file__), '../input_files', 'ESSOS_biot_savart_LandremanPaulQA.json')
+json_file = os.path.join(os.path.dirname(__name__), 'input_files', 'ESSOS_biot_savart_LandremanPaulQA.json')
```
Comment (rogeriojorge) [inline examples/particle_tracing/trace_particles_coils_guidingcenter.py]: Didn't we see that this kind of change breaks for other users? Do you really need it?

File: examples/particle_tracing/trace_particles_vmec_Electric_field.py [hunk 4]
```diff
@@ -24,11 +24,11 @@
 energy=FUSION_ALPHA_PARTICLE_ENERGY
 
 # Load coils and field
-wout_file = os.path.join(os.path.dirname(__file__), "../input_files", "wout_LandremanPaul2021_QA_reactorScale_lowres.nc")
+wout_file = os.path.join(os.path.dirname(__name__), "input_files", "wout_LandremanPaul2021_QA_reactorScale_lowres.nc")
```
Comment (rogeriojorge) [inline examples/particle_tracing/trace_particles_vmec_Electric_field.py]: Same

File: examples/particle_tracing/trace_particles_vmec_Electric_field.py [hunk 5]
```diff
@@ -24,11 +24,11 @@
...
 vmec = Vmec(wout_file)
 
 #Load electric field
-Er_file=os.path.join(os.path.dirname(__file__), '../input_files','Er.h5')
+Er_file=os.path.join(os.path.dirname(__name__), 'input_files','Er.h5')
```
Comment (rogeriojorge) [inline examples/particle_tracing/trace_particles_vmec_Electric_field.py]: Same

File: examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+def loss_particle_radial_drift(field, particles, timestep=1.e-8, maxtime=1e-5, num_steps=300, trace_tolerance=1e-5, model='GuidingCenterAdaptative',boundary=None):
+    particles.to_full_orbit(field)
+    tracing = Tracing(field=field, model=model, particles=particles, maxtime=maxtime,
+                      timestep=timestep,times_to_trace=num_steps, atol=trace_tolerance,rtol=trace_tolerance,boundary=boundary)
+    xyz = tracing.trajectories[:,:, :3]
+    R_axis=field.r_axis
+    Z_axis=field.z_axis
+    #Ideally here one would differentiate in time through diffrax !TODO
+    r_cross=jnp.sqrt(jnp.square(jnp.sqrt(jnp.square(xyz[:,0])+jnp.square(xyz[:,1]))-R_axis+1.e-12)+jnp.square(xyz[:,2]-Z_axis+1.e-12))
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py]: I thought xyz is (nparticles, ntimes, 3). Shouldn't this be xyz[:,:,0]?

File: examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+phi_array = jnp.linspace(0, 2*jnp.pi, NPARTICLES)
+initial_xyz=jnp.array([LARGE_R*jnp.cos(phi_array), LARGE_R*jnp.sin(phi_array), 0*phi_array]).T
+particles = Particles(initial_xyz=initial_xyz)
+
+
+
+
+
+
+
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py]: Many line breaks?

File: examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+# Particle optimization parameters
+# Optimization parameters
+NPARTICLES = number_of_processors_to_use*10
+MAXTIME_TRACING = 1e-4
+NUMBER_COILS_PER_HALF_FIELD_PERIOD = 3
+NUMBER_OF_FIELD_PERIODS = 2
+MODEL = 'GuidingCenterAdaptative'
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py:23]: Is this used later? Seems like it calls particles.to_full_orbit(field) inside the loss. What about importing/using the existing loss_particle_radial_drift() from objective_functions.py instead of redefining it locally?

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_stochastic.py [hunk 9]
```diff
@@ -2,177 +2,184 @@
...
 
-# Initialize VMEC field
-vmec = Vmec(os.path.join(os.path.dirname(__file__), '..', 'input_files',
-             'wout_LandremanPaul2021_QA_reactorScale_lowres.nc'),
-            ntheta=ntheta, nphi=nphi, range_torus='full torus')
+def perturbed_field_from_field(field, key, sampler):
+    coils = copy_coils_from_field(field)
+    base_key = jax.random.key(key)
+    split_keys = jax.random.split(base_key, 2)
+    perturb_curves_systematic(coils, sampler, key=split_keys[0])
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_stochastic.py]: This is not assigned anywhere? so systematic perturbations are ignored? Should it be coils = perturb_curves_systematic(...) before applying the statistical perturbations?

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
A couple of changes asked for

File: examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py (hunk 8 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py:23]: I kept the exmaples with independent functions on purpose, for people to see what the example is doing more clearly wihtout having to do to the code @rogeriojorge .

File: tests/test_objective_functions.py (hunk 1 above)
Comment (eduardolneto) [reply tests/test_objective_functions.py]: Done

File: tests/test_objective_functions.py (hunk 2 above)
Comment (eduardolneto) [reply tests/test_objective_functions.py:124]: Done

File: examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py (hunk 7 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py]: Fixed

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_stochastic.py (hunk 9 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_stochastic.py]: Done

File: examples/particle_tracing/trace_particles_vmec_Electric_field.py (hunk 5 above)
Comment (eduardolneto) [reply examples/particle_tracing/trace_particles_vmec_Electric_field.py]: Done

File: examples/particle_tracing/trace_particles_vmec_Electric_field.py (hunk 4 above)
Comment (eduardolneto) [reply examples/particle_tracing/trace_particles_vmec_Electric_field.py]: Done

File: examples/particle_tracing/trace_particles_coils_guidingcenter.py (hunk 3 above)
Comment (eduardolneto) [reply examples/particle_tracing/trace_particles_coils_guidingcenter.py]: Done

**Comment by eduardolneto** [comment]
The only example not working is with from_simsopt method. but I dopnt have a solution for this yet. I would merge this and solve that after. The change fo PR #43 works also on older diffrax version. @rogeriojorge @EstevaoMGomes

File: tests/test_objective_functions.py [hunk 10]
```diff
@@ -1,260 +1,275 @@
...
+        self.assertTrue(jnp.isfinite(objf.loss_BdotN(self.field, self.surface)))
+        self.assertTrue(jnp.isfinite(objf.loss_BdotN_constraint(self.field, self.surface)))
+
+    def test_copy_coils_from_field(self):
+        copied = objf.copy_coils_from_field(self.field)
+        self.assertIsInstance(copied, DummyCoils)
+
+    @patch("essos.objective_functions.BiotSavart", return_value=DummyField())
+    @patch("essos.objective_functions.perturb_curves_systematic", create=True)
+    @patch("essos.objective_functions.perturb_curves_statistic", create=True)
```
Comment (rogeriojorge) [inline tests/test_objective_functions.py]: I think this is masking the fact that in objective_functions perturb_curves_systematic and perturb_curves_statistic are not imported but still used, which leads to error. Why not use the API already imported there, in objective_functions?

coils = perturb_curves(
    coils,
    sampler,
    key=split_keys[0],
    perturbation_type="systematic",
)
coils = perturb_curves(
    coils,
    sampler,
    key=split_keys[1],
    perturbation_type="statistical",
)

File: essos/objective_functions.py [hunk 11]
```diff
@@ -203,7 +203,7 @@ def perturbed_field_from_field(field, key, sampler):
     coils = copy_coils_from_field(field)
     base_key = jax.random.key(key)
     split_keys = jax.random.split(base_key, 2)
-    perturb_curves_systematic(coils, sampler, key=split_keys[0])
+    coils = perturb_curves_systematic(coils, sampler, key=split_keys[0])
     coils = perturb_curves_statistic(coils, sampler, key=split_keys[1])
```
Comment (rogeriojorge) [inline essos/objective_functions.py:207]: See previous comment:

coils = perturb_curves(
    coils,
    sampler,
    key=split_keys[0],
    perturbation_type="systematic",
)
coils = perturb_curves(
    coils,
    sampler,
    key=split_keys[1],
    perturbation_type="statistical",
)

File: examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py (hunk 6 above)
Comment (rogeriojorge) [reply examples/coil_optimization/optimize_coils_particle_confinement_guidingcenter.py]: @eduardolneto

File: examples/particle_tracing_collisions/statistics_collisions_velocity_distributions_mu_Adaptative.py [hunk 12]
```diff
@@ -70,102 +73,144 @@
...
+subset_size = 10
+import numpy as np
+subset_indices = np.random.choice(len(trajectories), subset_size, replace=False)
 
-for i, trajectory in enumerate(trajectories):
-    ax2.plot(tracing.times, (tracing.energy[i]-tracing.energy[i,0])/tracing.energy[i,0], label=f'Particle {i+1}')      
-    ax3.plot(tracing.times, trajectory[:, 3]*SPEED_OF_LIGHT/jnp.sqrt(tracing.energy[i]/mass*2.), label=f'Particle {i+1}')    
+for i in subset_indices:
+    trajectory = trajectories[i]
+    ax1.plot(trajectory[:,0], trajectory[:,1], trajectory[:,2], label=f'Particle {i+1}')
```
Comment (rogeriojorge) [inline examples/particle_tracing_collisions/statistics_collisions_velocity_distributions_mu_Adaptative.py:100]: I think for this to work you need

ax1 = fig.add_subplot(221, projection="3d")

Comment (rogeriojorge) [reply examples/particle_tracing_collisions/statistics_collisions_velocity_distributions_mu_Adaptative.py:100]: The same in the other collision examples

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
A couple of small things, and then it's ready to merge

File: examples/particle_tracing_collisions/statistics_collisions_velocity_distributions_mu_Adaptative.py (hunk 12 above)
Comment (eduardolneto) [reply examples/particle_tracing_collisions/statistics_collisions_velocity_distributions_mu_Adaptative.py:100]: Fixed.

File: essos/objective_functions.py (hunk 11 above)
Comment (eduardolneto) [reply essos/objective_functions.py:207]: Fixed.

File: tests/test_objective_functions.py (hunk 10 above)
Comment (eduardolneto) [reply tests/test_objective_functions.py]: Fixed.
